File: test/sc-stub-outcome-overlay/src/main/scala/CustomContainers.scala
package containers

sealed trait Outcome[+T]
object Outcome {
  case class Success[+T](value: T) extends Outcome[T]
  case class Failure[+T](error: Any) extends Outcome[T]
}

class OutcomeServiceRt extends testpkg.pkg0.IBaboonServiceRt {
  def pure[L, R](value: R): Outcome[R] = Outcome.Success(value)
  def fail[L, R](error: L): Outcome[R] = Outcome.Failure(error)
  def leftMap[A, B, C](value: Outcome[B], f: A => C): Outcome[B] = value match {
    case Outcome.Failure(e) => Outcome.Failure(f(e.asInstanceOf[A]))
    case s: Outcome.Success[B] => s
  }
  def flatMap[A, B, C](value: Outcome[B], f: B => Outcome[C]): Outcome[C] = value match {
    case Outcome.Failure(e) => Outcome.Failure(e)
    case Outcome.Success(v) => f(v)
  }
}

object OutcomeServiceRt {
  val instance: OutcomeServiceRt = new OutcomeServiceRt()
}

File: test/rs-stub-outcome-overlay/src/custom_containers.rs
use crate::testpkg::pkg0::baboon_service_rt::IBaboonServiceRt;

pub enum Outcome<T> {
    Success(T),
    Failure(Box<dyn std::any::Any>),
}

pub struct OutcomeServiceRt;

impl IBaboonServiceRt for OutcomeServiceRt {
    fn pure<L: 'static, R>(&self, value: R) -> Outcome<R> {
        Outcome::Success(value)
    }
    fn fail<L: 'static, R>(&self, error: L) -> Outcome<R> {
        Outcome::Failure(Box::new(error))
    }
    fn left_map<A: 'static, B, C: 'static, F: FnOnce(A) -> C>(&self, value: Outcome<B>, f: F) -> Outcome<B> {
        match value {
            Outcome::Failure(e) => {
                if let Ok(a) = e.downcast::<A>() {
                    Outcome::Failure(Box::new(f(*a)))
                } else {
                    panic!("type mismatch in left_map");
                }
            }
            Outcome::Success(v) => Outcome::Success(v),
        }
    }
    fn flat_map<A: 'static, B, C, F: FnOnce(B) -> Outcome<C>>(&self, value: Outcome<B>, f: F) -> Outcome<C> {
        match value {
            Outcome::Failure(e) => Outcome::Failure(e),
            Outcome::Success(v) => f(v),
        }
    }
}
